### ISSUE TITLE
"OSError: [Errno 49] Can't assign requested address" for IPv6 address

### ISSUE THREAD

**Issue by adam-pawluczuk** [agronholm/anyio#554]
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### AnyIO version

3.6.2

### Python version

3.10.6

### What happened?

`anyio.create_tcp_listener` was not able to bind to `fe80::1%lo0`.

### Details

It looks like the `create_tcp_listener` always resolves IPv6 addresses to a 2-element tuples:
* https://github.com/agronholm/anyio/blob/3.6.2/src/anyio/_core/_sockets.py#L500

Though 2-element tuples for IPv6 addresses are not always understandable by `socket.bind()`.

From https://docs.python.org/3/library/socket.html#socket-families:

> For [AF_INET6](https://docs.python.org/3/library/socket.html#socket.AF_INET6) address family, a four-tuple (host, port, flowinfo, scope_id) is used, where flowinfo and scope_id represent the sin6_flowinfo and sin6_scope_id members in struct sockaddr_in6 in C. For [socket](https://docs.python.org/3/library/socket.html#module-socket) module methods, flowinfo and scope_id can be omitted just for backward compatibility. Note, however, omission of scope_id can cause problems in manipulating scoped IPv6 addresses

`anyio.create_tcp_listener` under the hood uses `anyio.getaddrinfo`. 

It looks like the way how `anyio.getaddrinfo` works for IPv6 addresses is not fully compatible with the way how `socket.bind` is expecting the IPv6 addresses format to be.

`anyio.getaddrinfo` parses `fe80::1%lo0` (which is a link-local address) as `('fe80::1%1', 0)` (where `%1` is `scope_id`), whereas `socket.bind` is expecting `scope_id` as a separate parameter.

This is what `anyio.getaddrinfo` produces:
* `('fe80::1%1', 0)`

For comparison - this is the result of using `socket.getaddrinfo`:
```python
>>> import socket
>>> socket.getaddrinfo("fe80::1%lo0", 0, proto=socket.IPPROTO_TCP)
[(<AddressFamily.AF_INET6: 30>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('fe80::1', 0, 0, 1))]
```

So we have `('fe80::1%1', 0)` vs. `('fe80::1', 0, 0, 1)`.

I'm not able to bind to `('fe80::1%1', 0)` using:
```python
import socket
raw_socket = socket.socket(socket.AddressFamily.AF_INET6)
raw_socket.bind(('fe80::1%1', 0))
```

This is what `socket.bind` expects (both work):
* `('fe80::1%1', 0, 0, 1)`
* `('fe80::1', 0, 0, 1)`

I am able to bind:
```python
import socket
raw_socket = socket.socket(socket.AddressFamily.AF_INET6)
raw_socket.bind(('fe80::1%1', 0, 0, 1))
# or
# raw_socket.bind(('fe80::1', 0, 0, 1))
```

### How can we reproduce the bug?

1. Add this line to `/etc/hosts`:
    * `fe80::1%lo0     localhost`
2. Try to use `anyio.create_tcp_listener(host="localhost", port=0)`.
3. Which results in:
    * `OSError: [Errno 49] Can't assign requested address` 

Or even simpler:
1.  `anyio.create_tcp_listener(host="fe80::1%lo0", port=0)`

**Comment by agronholm** [comment]
Ok, so I have a question: where does `fe80::1%lo0` actually come from? The IPv6 address for `localhost` is `::1`, not `fe80::1`.

**Comment by jonathanslenders** [comment]
This seems to be the link-local address for the loopback interface. It can be like that in /etc/hosts and localhost can resolve to this address.
See: https://superuser.com/questions/241642/what-is-the-relevance-of-fe801lo0-localhost-in-etc-hosts

**Comment by agronholm** [comment]
What OS is `fe80::1%lo0` a valid interface on? I tried on all of mine, and the result was always the same: `socket.gaierror: [Errno -2] Name or service not known`.

**Comment by smurfix** [comment]
I have no idea which OS variant actually adds this address without explicitly being told to, but on Linux you can easily do an `ip -6 addr add fe80::1/128 dev lo` if you want to test this.

Depends on the name of your loopback interface of course.

**Comment by adam-pawluczuk** [comment]
> What OS is fe80::1%lo0 a valid interface on?

I'm on Mac.

```
Darwin [...] 22.4.0 Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:17 PST 2023; root:xnu-8796.101.5~3/RELEASE_X86_64 x86_64
```

`getaddrinfo` is able to resolve:

```pycon
>>> import socket
>>> socket.getaddrinfo("fe80::1%lo0", 80, family=socket.AddressFamily.AF_INET6, proto=socket.IPPROTO_TCP)
[(<AddressFamily.AF_INET6: 30>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('fe80::1', 80, 0, 1))]
```

And then I can also bind:
```pycon
>>> s.bind(("fe80::1", 0, 0, 1))
>>> 
```

>>> s = socket.socket(socket.AddressFamily.AF_INET6)
> I have no idea which OS variant adds this address without explicitly being told to

The issue I was having was that added this address explicitly (not that it was added by the OS itself).

**Comment by agronholm** [comment]
This is starting to sound like a made-up problem. The AnyIO test suite does test binding to the `localhost` interface on IPv6, without issues so far. The reason why AnyIO's `getaddrinfo()` returns a 2-tuple always is for consistently being able to unpack to (`host`, `port`) regardless of IP protocol version. I'd need a lot more convincing to break the existing API, and it would have to be done for v4.0 which should be out within the next month.

**Comment by agronholm** [comment]
The problem here seems to be that the name resolution results in an address which _itself_ needs to be resolved to an actionable socket address. Why it's not getting one already from the backend's `getaddrinfo()` call is a bit of a mystery. Is it just naively returning the address from `/etc/hosts`?

**Comment by smurfix** [comment]
Well, link-local addresses are a thing in the IPv6 world (and IPv4 too if you enable them). So you definitely need to be able to pass the scope ID, a.k.a. interface number, through the stack when you bind a socket to a link-local address. The host name by itself isn't sufficient because the address may well be valid on more than one interface.

There are protocols for which this is strongly relevant; Avahi/Zeroconf clients habitually use link-local addresses, and I might want to use AnyIO to write one that doesn't go splat when the box it's running on has multiple interfaces using them, which apparently includes any Mac, as they enable them on loopback. This might actually simplify the Avahi server's code (no need to special-case `lo0`), so it's not _quite_ as silly as I initially thought.

> Why it's not getting one already from the backend's getaddrinfo() call is a bit of a mystery.

The Python docs state that starting with Py3.7 the address returned by `getaddrinfo` (i.e. the tuple's first member) no longer contains a `%scope_id` trailer.

> 4-tuple IPv6 socket addresses are automatically converted to 2-tuples of (host, port), unlike what [socket.getaddrinfo()](https://docs.python.org/3/library/socket.html#socket.getaddrinfo) does.

(from anyio3's documentation) 

Well, starting with Py3.7 this results in loss of information which you need to actually `bind` or `connect` to the address in question. Thus I somewhat-strongly recommend not to do that. Just pass the system's result tuple through as-is.

**Comment by agronholm** [comment]
> Well, link-local addresses are a thing in the IPv6 world (and IPv4 too if you enable them). 

I am not arguing against the usefulness of binding to link-local addresses.

> So you definitely need to be able to pass the scope ID, a.k.a. interface number, through the stack when you bind a socket to a link-local address.

Yes, that's what the suffix `%1` or whatever is for.

> The Python docs state that starting with Py3.7 the address returned by getaddrinfo (i.e. the tuple's first member) no longer contains a %scope_id trailer.

I know, that's why AnyIO's `getaddrinfo()` uses the `convert_ipv6_sockaddr()` function to append it to the host part.

> Well, starting with Py3.7 this results in loss of information which you need to actually bind or connect to the address in question. Thus I somewhat-strongly recommend not to do that. Just pass the system's result tuple through as-is.

What information is lost? The scope ID is just moved from the last tuple element to the host address.

**Comment by agronholm** [comment]
Before I do anything about this one way or another, I'm going to run some tests and make sure I understand everything correctly.

**Comment by agronholm** [comment]
I've looked into this further and it seems that I indeed misunderstood how `socket.bind()` works with the scope ID. Sorry for dismissing this earlier! I thought that a call like `sock.bind(("fe80::1%1", 0))` would be valid, but it's really not (gives `Invalid argument`). The reason for my confusion is that for anything else than link-local addresses, `bind()` works just fine with a 2-tuple address, even for IPv6, and I thought that it would accept an IPv6 address with the proper scope ID in it that way, but seems that it doesn't. For now, I will work around this in v3.7.0, and change the API for v4.0. Testing this in CI might be a problem though.

**Comment by agronholm** [comment]
Curiously, only `create_tcp_listener()` seems to be affected. I tested the following socket functions I thought might be broken, but they all worked fine:
* `connect_tcp()`  (using the link-local addr as the remote address)
* `create_connected_udp_socket()` (using the link-local addr as the remote address)
* create_udp_socket()

**Comment by agronholm** [comment]
I have a working test for this now, but I discovered that it fails with `uvloop`, as the returned scope ID is incorrect. Works fine on vanilla asyncio and trio.

**Comment by agronholm** [comment]
I found what triggers this: uvloop seems to behave a bit oddly with `type=SocketKind.SOCK_STREAM`. The returned scope ID is correct if I omit this parameter. No clue why this happens. I _think_ I can safely omit this argument though.

**Comment by agronholm** [comment]
Minimum workable example (requires `psutil`):
```python
from asyncio import get_running_loop, run
from socket import SocketKind

import psutil
import uvloop


async def main():
    for ifname, addresses in psutil.net_if_addrs().items():
        for addr in addresses:
            if addr.address.startswith("fe80::"):
                gai_res = await get_running_loop().getaddrinfo(addr.address, 0, type=SocketKind.SOCK_STREAM)
                # Ensure that the returned scope ID is not 0
                assert gai_res[-1][-1][-1] != 0
                return

    raise Exception("Could not find a suitable link-local address")

uvloop.install()
run(main())
```

**Comment by agronholm** [comment]
@adam-pawluczuk Would you mind testing against the linked PR?

**Comment by smurfix** [comment]
> I tested the following socket functions I thought might be broken, but they all worked fine:

They might not work fine when the remote end you connect to is on the "wrong" interface. I.e. when you have two non-`lo` interfaces with link-local addresses on them, and need to use a specific interface to connect to a given remote address.

**Comment by agronholm** [comment]
You would have to give that interface name in the address after the `%` suffix then. I think you would get an `OSError` if you don't.

On another note, my workaround for the uvloop bug doesn't work on Windows, as it disregards `type=SocketKind.SOCKSTREAM` there, so the check for the socket kind fails, and no addresses are selected.

**Comment by agronholm** [comment]
Side note: I discovered (and worked around) a bug in PyPy's implementation of `socket.getaddrinfo()`. Once I [reported it](https://foss.heptapod.net/pypy/pypy/-/issues/3938), it was promptly fixed within 2 hours. Kudos to the PyPy folks for the lightning speed fix!

**Comment by agronholm** [comment]
Fixed in 785f697d.
